pulsar-client/src/test/java/org/apache/pulsar/client/impl/conf/ClientConfigurationDataTest.java

```diff
@@ -18,38 +18,102 @@
+import org.apache.pulsar.client.impl.auth.AuthenticationDisabled;
 import org.apache.pulsar.client.impl.auth.AuthenticationToken;
-import org.testng.Assert;
 import org.testng.annotations.Test;
 
 /**
- * Unit test {@link ClientConfigurationData}.
+ * Unit tests for {@link ClientConfigurationData}.
  */
 public class ClientConfigurationDataTest {
 
-    private final ObjectWriter w;
-
-    {
-        ObjectMapper m = new ObjectMapper();
-        m.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
-        w = m.writer();
-    }
-
+    private static final String BASIC_AUTH_PLUGIN_CLASS = AuthenticationBasic.class.getName();
 
     @Test
     public void testDoNotPrintSensitiveInfo() throws JsonProcessingException {
         ClientConfigurationData clientConfigurationData = new ClientConfigurationData();
         clientConfigurationData.setTlsTrustStorePassword("xxxx");
         clientConfigurationData.setSocks5ProxyPassword("yyyy");
         clientConfigurationData.setAuthentication(new AuthenticationToken("zzzz"));
-        String s = w.writeValueAsString(clientConfigurationData);
-        Assert.assertFalse(s.contains("xxxx"));
-        Assert.assertFalse(s.contains("yyyy"));
-        Assert.assertFalse(s.contains("zzzz"));
+        ObjectMapper objectMapper = new ObjectMapper();
+        objectMapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
+        String serializedConf = objectMapper.writeValueAsString(clientConfigurationData);
+        assertThat(serializedConf).doesNotContain("xxxx", "yyyy", "zzzz");
+    }
+
+    @Test
+    public void testGetAuthWhenSetFromPropsUsingAuthParams() {
+        ClientConfigurationData conf = new ClientConfigurationData();
+        conf.setAuthPluginClassName(BASIC_AUTH_PLUGIN_CLASS);
+        conf.setAuthParams(basicAuthParams("user1", "pass1"));
+        assertThatConfHasBasicAuth(conf, "user1", "pass1");
+    }
+
+    @Test
+    public void testGetAuthWhenSetFromPropsUsingAuthParamMap() {
+        ClientConfigurationData conf = new ClientConfigurationData();
+        conf.setAuthPluginClassName(BASIC_AUTH_PLUGIN_CLASS);
+        conf.setAuthParamMap(basicAuthParamMap("user1", "pass1"));
+        assertThatConfHasBasicAuth(conf, "user1", "pass1");
+    }
+
+    @Test
+    public void testGetAuthWhenSetFromPropsWithAllPropsAvailable() {
+        ClientConfigurationData conf = new ClientConfigurationData();
+        conf.setAuthPluginClassName(BASIC_AUTH_PLUGIN_CLASS);
+        conf.setAuthParams(basicAuthParams("user1", "pass1"));
+        conf.setAuthParamMap(basicAuthParamMap("user2", "pass2"));
+        assertThatConfHasBasicAuth(conf, "user1", "pass1");
+    }
+
+    private void assertThatConfHasBasicAuth(ClientConfigurationData conf, String userId, String password) {
+        assertThat(conf.getAuthentication())
+                .isInstanceOf(AuthenticationBasic.class)
+                .hasFieldOrPropertyWithValue("userId", userId)
```

Comment:
Don't suggest you use the `hasFieldOrPropertyWithValue` to assert.  

I suggest you add the following methods to the `AuthenticationBasic.class`:
```
    @VisibleForTesting
    public String getUserId() {
        return userId;
    }

    @VisibleForTesting
    public String getPassword() {
        return password;
    }
```

And call a method to assert.

Reply:
Is the concern that using the by-field-name assertion makes it hard to "find usages" when refactoring later? 

One concern I have is exposing the password w/ a public getter. However, I do see that other auth classes have taken a similar approach (eg [AuthenticationTls](https://github.com/apache/pulsar/blob/06af2ccf1fe9f0f0a89eb79659915e5801da9dd8/pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationTls.java#L128)).

Reply:
> Is the concern that using the by-field-name assertion makes it hard to "find usages" when refactoring later?

Right.

> One concern I have is exposing the password w/ a public getter. However, I do see that other auth classes have taken a similar approach (eg [AuthenticationTls](https://github.com/apache/pulsar/blob/06af2ccf1fe9f0f0a89eb79659915e5801da9dd8/pulsar-client/src/main/java/org/apache/pulsar/client/impl/auth/AuthenticationTls.java#L128)).

In my opinion, I can accept this approach, or you can mock an auth to test.

Reply:
@nodece the latest commit uses a local mock auth instead of basic auth. Its a bit more code but I think its worth not relying on one of the auth impls as well as not having to expose the fields of one of the auth impls. Lmk what you think. If you want to go back to basic auth I can revert the commit.

Reply:
Right now, LGTM.